Integrating the Instagram Django Work load in Cloudsuite

jayadevgithub opened the pull request:
These changes are related to the Instagram Django workload integration into the cloudsuite.

**Note**: Django workload uses Memcached and Cassandra servers. As we already have these servers as a part of our existing data-caching and data-serving benchmark applications, we have made necessary changes to these existing servers and clients to make them compatible to the Django workload also. These common server files are now moved under **cloudsuite/commons** directory.

The following are the changes made:
1. Cassandra is now common to both Data Serving(server) and Django workloads and is found in cloudsuite/commons/cassandra with necessary client changes made.
**Note**: To get the cassandra server up and running in the Data Serving Application use the following updated docker run command (update is, **--privileged** is added):
`docker run --privileged --name cassandra-server --net serving_network cloudsuite/data-serving:server cassandra`
2. Memcached is now common compatible to both Data Caching(server) and Django workloads and is found in cloudsuite/commons/memcached with necessary changes made.
 **Note:** There is no change in the run commands to set up the Data Caching client and server, the exisitng documentation can be followed.
3. To set up the Instagram/Django workload, please refer to the **README.md** within the django-workload project.

**TODO**:

1. Make changes to the git urls to point to the parsa-epfl/cloudsuite in the siege and uwsgi Dockerfiles of the django-workload once these changes are approved.
2. Delete the **server** directories under data-caching & data-serving benchmarks upon approval.
3. Update the cloudsuite documentation for the data-serving, data-caching and add new documentation for django-workload.

jayadevgithub committed:
initial commit for instagram django workload

Commit:
memcached changes in data-caching and django

jayadevgithub committed:
temp commit

jayadevgithub committed:
Merge branch 'master' into django_workload

jayadevgithub committed:
cassandra client changes

jayadevgithub committed:
refactored codebase, added memcached and cassandra into commons folder

jayadevgithub committed:
Merge pull request #2 from parsa-epfl/master

uptaking changes from cloudsuite project

jayadevgithub committed:
updated readme

varun729 commented:
A few problems as I try to setup the workload for the first time:
_**[Jayadev]**: Resolved the issues. Please find resolution specific comments inline._

1. The description of the workload is missing. For other cloudsuite workloads, the description is found in the cloudsuite.ch website, so in the future we might move the description there. But, for now I don't understand the purpose of the docker containers.
_**[Jayadev]**: Added the description to a README.md_

2. Cassandra and memcache containers are moved to the commons directory, along with the files that are used by the workload. Containers in the commons directory are supposed to be shared, so I think a stripped down version of the cassandra and memcache containers should be in the commons directory, and the workload specific container should use the common images. For example, for cassandra we can move the jvm.options.128GB file to django workload directory, and build a workload specific docker image for cassandra.
_**[Jayadev]**: Made the changes as requested._

3. Instructions for building uWSGI are confusing - there are two ways to build the uWSGI docker image. We should figure out the one that is best and stick to it.
_**[Jayadev]**: Made the instructions clear, we can still discuss if we want proceed with a single implementation._

4. Starting the cassandra container takes a long time. If possible perform some of the tasks while building the docker image.
_**[Jayadev]**: Figured out why this happens. It's due to the jvm settings. You might have to configure the jvm memory parameters based on the system. I have mentioned that in the README.md_

5. If possible, we should do the setup without using docker network, so that it is easier to run the containers on different hosts.
_**[Jayadev]**: Made the necessary changes._

6. Running the benchmark, as of now, takes a long time (more than 30 minutes, it is still running). The workload runs multiple iterations, so there should be a way to pass a command line option limiting the number of iterations. I didn't find the command line option in either the Dockerfile, entrypoint script or the run_containers.sh script.
_**[Jayadev]**: I have run multiplte times now and the benchmark consistenlty ends after 7 Iterations. I will figure out if there is a way to limit these iterations._

7. The instructions do not explain what will be the result of running the workload. So, I don't know what to expect once the workload ends.
_**[Jayadev]**:  After the benchmark is run successfully, we get the following summary:_

```
Running with 185 Siege workers
Running iteration 1 --- DONE
WARNING: Got 116 socket timeout alerts
Please see full Siege log in /tmp/siege_out_1

Running iteration 2 --- DONE
WARNING: Got 110 socket timeout alerts
Please see full Siege log in /tmp/siege_out_2

Running iteration 3 --- DONE
WARNING: Got 115 socket timeout alerts
Please see full Siege log in /tmp/siege_out_3

Running iteration 4 --- DONE
WARNING: Got 119 socket timeout alerts
Please see full Siege log in /tmp/siege_out_4

Running iteration 5 --- DONE
WARNING: Got 92 socket timeout alerts
Please see full Siege log in /tmp/siege_out_5

Running iteration 6 --- DONE
WARNING: Got 142 socket timeout alerts
Please see full Siege log in /tmp/siege_out_6

Running iteration 7 --- DONE
WARNING: Got 119 socket timeout alerts
Please see full Siege log in /tmp/siege_out_7


Removing results with Transaction rate min=191.74 and max=205.53

URL hit percentages:
/timeline:              25.5560052871%, expected 25%
/inbox:                 20.0321742387%, expected 19%
/feed_timeline:         24.9417829531%, expected 26%
/bundle_tray:           22.9094493075%, expected 25%
/seen:                  6.55975300102%, expected 5%

Transactions:                           23623.2 hits ---- RSD 0.0238203033983
Availability:                           99.516 % ---- RSD 0.000526379711954
Elapsed time:                           119.734 secs ---- RSD 0.00302359917196
Data transferred:                       46.624 MB ---- RSD 0.0246023260051
Response time:                          0.766 secs ---- RSD 0.0177084333763
Transaction rate:                       197.292 trans/sec ---- RSD 0.0227308983587
Throughput:                             0.388 MB/sec ---- RSD 0.0252524715751
Concurrency:                            151.142  ---- RSD 0.018205388363
Successful transactions:                23623.4  ---- RSD 0.0238249789264
Failed transactions:                    114.6  ---- RSD 0.0931028179291
Longest transaction:                    28.894  ---- RSD 0.0300870553859
Shortest transaction:                   0.11  ---- RSD 0.572077553547
P50:                                    0.542 secs ---- RSD 0.0215164276563
P90:                                    0.746 secs ---- RSD 0.0648992945679
P95:                                    1.564 secs ---- RSD 0.0103887959139
P99:                                    7.092 secs ---- RSD 0.110873942728
```

varun729 commented:
The workload has still not finished after 11 hours. I see the following
_**[Jayadev]**: This should get resolved in the latest code._ 
> ...
> Starting siege container
> Running with 185 Siege workers
> Running iteration 1 --- DONE
> WARNING: Got 12 socket timeout alerts
> Please see full Siege log in /tmp/siege_out_1
> 
> Running iteration 2
> 
>

jayadevgithub committed:
refactored commons/cassandra and benchmarks/data-serving files

jayadevgithub committed:
refactored commons/memcached and benchmarks/data-caching files

jayadevgithub committed:
refactored benchmarks/django-workload files

jayadevgithub committed:
more refactoring to the django-workload: README files, build & run scripts for each component, removed docker network dependency

jayadevgithub commented:
@varun729 : Addressed your comments that you mentioned [here](https://github.com/jayadevgithub/cloudsuite/pull/3#issuecomment-422669423). Please find the responses inline.

varun729 commented:
What are the "WARNING: Got 119 socket timeout alerts" messages?

varun729 commented:
The filename "jvm.options.128GB" is inappropriate, it suggests something should be 128GB.

You should figure out a formula to compute the values for Xmx, Xms, etc for a given amount of memory, which the user can define on the command line when the user enters "docker run" command. The bootstrap.sh file can then use the parameter to update these JVM options.

Based on the current setup, we need different docker images for different sizes of memory. We just build one docker image, and it should be parameterized.

_**[Jayadev]:** Made changes as suggested_

varun729 commented:
In commons/cassandra, I see that you are installing oracle-java8. Is there a reason you need Oracle Java and not OpenJDK. Cloudsuite already has docker images for OpenJDK, which I think you can use for cassandra.

_**[Jayadev]:** Made changes as suggested_

varun729 commented:
For both commons/cassandra and commons/memcache you are installing nf_conntrack to /etc/modules.
What is nf_conntrack?

_**[Jayadev]:** Looks like **nf_conntrack** is from the set of the **conntrack-tools** that apparantly allows system administrators interact, from user-space, with the in-kernel Connection Tracking System, which is the module that enables stateful packet inspection for iptables. We haven't used this earlier, and I don't see the usage of this or even the need to use this now. So I got rid of this module in the latest commit._

jayadevgithub committed:
refactored commons/memcache to include a version based installation followe by changes to data-caching/server, django-workload/memcached

jayadevgithub committed:
configured cassandra bootstrap scripts to set jvm options dynamically based on the system memory which is given as an env variable while executing docker run.
Example: docker run -e SYSTEM_MEMORY=128...

jayadevgithub committed:
Using openjdk8 from cloudsuite/java instead on a fresh installation. Removed adding nf_conntrack to etc/modules for cassandra image as we haven't used earlier and we aren't using it for any connection tracking explicitly in our application
